src/cli/self_update/shell.rs

```diff
@@ -233,7 +233,8 @@ impl UnixShell for Fish {
     }
 
     fn update_rcs(&self) -> Vec<PathBuf> {
-        self.rcfiles()
+        // The first rcfile takes precedence.
```

Comment:
This doesn't obviously make sense to me. Why is this necessary? At the very least deserves some more comments on why we're doing this.

Reply:
@djc Sure! The snippet below explains the differences between `.rcfiles()` and `.update_rcs()`.

https://github.com/rust-lang/rustup/blob/3fcd56283311ab1cc24ab275bc7e44b9d7f62a8d/src/cli/self_update/shell.rs#L91-L96

The problem with the current code is that, we have two rc paths in fish, one (`p1`) being the **_fallback_** of the other (`p0`).

https://github.com/rust-lang/rustup/blob/3fcd56283311ab1cc24ab275bc7e44b9d7f62a8d/src/cli/self_update/shell.rs#L221-L230

According to the current implementation, if both paths are valid, then we'll create and append to both of them.

https://github.com/rust-lang/rustup/blob/f6dbe2e4cc186c7624e76691ece2ff37c87cf565/src/cli/self_update/unix.rs#L88-L93

This is the expected behavior in some cases, but clearly not for zsh or fish due to this fallback relationship.

https://github.com/rust-lang/rustup/blob/f6dbe2e4cc186c7624e76691ece2ff37c87cf565/src/cli/self_update/shell.rs#L193-L206

Reply:
So should we change `update_rcs()` to yield `PathBuf` instead of `Vec<PathBuf>`? It seems conceptually confusing to do this only for some shells -- do we update all rcs for `Bash`?

Anyway, I guess this change is fine, but let's add some comments to explain the logic.

This way of writing it avoids the `cloned()`, and thus seems cleaner to me:

```rust
match self.rcfiles().into_iter().next() {
    Some(path) => vec![path],
    None => vec![],
}
```

What do you think?

Reply:
> This way of writing it avoids the `cloned()`, and thus seems cleaner to me:
> 
> ```rust
> match self.rcfiles().into_iter().next() {
>     Some(path) => vec![path],
>     None => vec![],
> }
> ```
> 
> What do you think?

@djc As you wish.

Reply:
> This way of writing it avoids the `cloned()`, and thus seems cleaner to me:
> 
> ```rust
> match self.rcfiles().into_iter().next() {
>     Some(path) => vec![path],
>     None => vec![],
> }
> ```
> 
> What do you think?

@djc Done.

Reply:
Would still like some more comments about why we're doing this here.

Reply:
> https://github.com/rust-lang/rustup/blob/f6dbe2e4cc186c7624e76691ece2ff37c87cf565/src/cli/self_update/shell.rs#L138-L152
> 
> ... a quick blame points to #2387. @workingjubilee maybe you can provide more context on this one?

My original implementation was very cautious to avoid writing new files to the system, so as to leave the system maximally undisturbed. The problem was that people would then use shells that had not generated even stub rcfiles for themselves. This was most prominent with zsh, as I recall, where macOS, last I checked, generates neither a zshrc nor a zprofile. So what you are seeing is that original implementation, before it was paved over with our current reality.

Reply:
We also originally did not handle fish at all.

Reply:
@workingjubilee Thanks for your quick response! Apart from what you said above, we are particularly curious about why `update_rcs` is plural 👀

Reply:
> We also originally did not handle fish at all.

... and that's exactly why we need this context to evaluate our fish support implementation :]

Reply:
Why would it be singular?

Reply:
It was just my intuition that it wouldn't make sense to "update" multiple RC files per shell? It seems like if we want to make the rustup proxies work in fish, it would be enough if we update one file? So in that sense `update_rcs() -> Vec<PathBuf>` should be more like `update_rc() -> PathBuf`?

(I'm not big on shell customization so my understanding might be wrong.)

Reply:
@djc While imprecise on technical terms, my remarks in https://github.com/rust-lang/rustup/issues/3256#issuecomment-1462306973 remain an accurate summary of the actual situation. In short: No. Not at all. That requires knowledge of not only the OS, not only the shell (and we don't actually know that), but the specific version and distribution of the software.